### PR TITLE
Migrate legacy import logic to adapters and hide Import page 2

### LINKED ISSUE #11753: Bug in the page Import
**Describe the bug**
In the BO => Advanced Parameters => Import page, an exception is displayed

**To Reproduce**
1. Go to the BO => Advanced Parameters => Import page

**Screenshots**
capture d ecran de 2018-12-12 17-36-03

**Additionnal information**
PrestaShop version: develop
PHP version: 7.1

### PR DESCRIPTION
**Pull request by rokaszygmantas**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Legacy import logic migration to adapters, in order to free and remove legacy import controller.
| Type?         | improvement
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | https://github.com/PrestaShop/PrestaShop/issues/11753
| How to test?  | Import should work the same as before for ~~all import types~~ products and categories import types. A good way to test is to import the same csv file with same settings on two copies of prestashop (one with this PR and other with stable release), and then compare if the imported data is identical. It should also fix https://github.com/PrestaShop/PrestaShop/issues/11753. **Please build the assets when testing this PR.**

P.S. Noticed that when you have multistore enabled and "All shops" selected, you cannot import products - stock update fails. You must select a single shop to import it.

My internal checklist for this PR:
- [x] Product import logic moved to adapters
- [x] Category import logic moved to adapters

Futher issues to be solved in additional PRs:
- [ ] Combinations import logic moved to adapters
- [ ] Customers import logic moved to adapters
- [ ] Addresses import logic moved to adapters
- [ ] Manufacturers import logic moved to adapters
- [ ] Suppliers import logic moved to adapters
- [ ] Aliases import logic moved to adapters
- [ ] Store contacts import logic moved to adapters
- [ ] Supply orders import logic moved to adapters
- [ ] Supply orders details import logic moved to adapters
- [ ] Check this PR: https://github.com/PrestaShop/PrestaShop/pull/11715
- [ ] Unit tests
- [ ] Integration tests
- [ ] Run code sniffer
- [x] Write doc about internal and external behavior of the import process

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/11561)

### CODE REVIEW

**Comment by rokaszygmantas** [comment]
Hi @matks, 

This is one of the two PR discussed here: https://github.com/PrestaShop/PrestaShop/pull/11160#issuecomment-433317318
the goal is to migrate the import logic away from legacy controller into adapters.

At the moment in this PR only product import logic is migrated, but I'm working on migrating other import types as well.
A review would be handy, so I could fix the issues before proceeding with other import types :slightly_smiling_face:

**Comment by rokaszygmantas** [comment]
Btw the PR is huge, but mostly because of the amount of logic in one adapter (`ProductImportHandler`), I didn't change most of it, just copied over from the legacy controller, with some minor refactoring :slightly_smiling_face:

**Comment by matks** [comment]
@rokaszygmantas nice ! I will try to get a review today or tomorrow 😄

File: admin-dev/themes/new-theme/js/pages/import-data/Importer.js
Comment (matks) [inline admin-dev/themes/new-theme/js/pages/import-data/Importer.js:86]: 👍

File: admin-dev/themes/new-theme/js/pages/import-data/Importer.js
Comment (matks) [inline admin-dev/themes/new-theme/js/pages/import-data/Importer.js]: Is it possible to explain / make more readable (comment or dedicated function) this code statement ? I see the logic but cannot figure out why you dot it 🤔

File: src/Adapter/Converter/ExcelToCsvFileConverter.php
Comment (matks) [inline src/Adapter/Converter/ExcelToCsvFileConverter.php]: This looked good though 😛

File: src/PrestaShopBundle/Resources/config/services/adapter/data_provider.yml
Comment (matks) [inline src/PrestaShopBundle/Resources/config/services/adapter/data_provider.yml:112]: What happens if `getContext()->employee` returns null ? I'm thinking about Webservice usecase for example.

We are only migrating the BO right now so it will not be an issue for now however I would like us to check what happens for a not-BO usecase (where you might not have an employee) and handle it with a nice exception 😄

File: src/Adapter/Import/CsvFileOpener.php
Comment (matks) [inline src/Adapter/Import/CsvFileOpener.php:94]: Does calling `rewing($handle)` a 2nd time really makes a difference 🤔 ?

File: src/Adapter/Import/Handler/AbstractImportHandler.php
Comment (matks) [inline src/Adapter/Import/Handler/AbstractImportHandler.php:237]: Is that logic from the legacy codebase ?

**Comment by matks** [comment]
The internal list looks good 👍 can you add a step "write doc about internal and external behavior of the import process" ? This is necessary IMO 😉

File: src/Adapter/Import/Handler/AbstractImportHandler.php
Comment (matks) [inline src/Adapter/Import/Handler/AbstractImportHandler.php]: Can we make this more readable by using longer variable names ?

File: src/Adapter/Import/Handler/AbstractImportHandler.php
Comment (matks) [inline src/Adapter/Import/Handler/AbstractImportHandler.php]: Maybe we can use variable names as hints for readability, for example `$canBeTranslated` ?

```
$canBeTranslated = (in_array($field, $this->languageFields) && $languageId);
if ($canBeTranslated) {
    ...
}
```

File: src/Core/Import/Handler/ImportHandlerInterface.php
Comment (matks) [inline src/Core/Import/Handler/ImportHandlerInterface.php:45]: Should it really be part of the interface or part of the abstract ? Do all imports need `setUp()` ? Same for `tearDown()`

File: src/Adapter/Import/Handler/CategoryImportHandler.php
Comment (matks) [inline src/Adapter/Import/Handler/CategoryImportHandler.php:386]: I know this comes from the legacy codebase and it took you a lot of efforts to extract it. However I have to ask: doesn't it look strange to have this here ? Import logic that creates a category following its own rules ? Maybe we should have a single service where we handle the logic required to create a category (so same rules apply, whether we create a category from the BO page, from the Import page or using the webservice) ?

Comment (matks) [reply src/Adapter/Import/Handler/CategoryImportHandler.php:386]: Note to myself: bad idea 😅 

(I looked at how legacy category controller works)

File: src/Adapter/Import/Handler/CategoryImportHandler.php
Comment (matks) [inline src/Adapter/Import/Handler/CategoryImportHandler.php:52]: So you started other handlers as well, more than product 😉 ?

File: src/Adapter/Import/Handler/ProductImportHandler.php
Comment (matks) [inline src/Adapter/Import/Handler/ProductImportHandler.php:327]: My heart bleeds 💔

File: src/Adapter/Import/Handler/ImportHandlerFinder.php
Comment (matks) [inline src/Adapter/Import/Handler/ImportHandlerFinder.php]: You could add a `supports()` function to ImportHandlers to check whether an import handles can handle a type. Then you do not depend on an array key

File: src/Adapter/Import/ImportEntityDeleter.php
Comment (matks) [inline src/Adapter/Import/ImportEntityDeleter.php:176]: Why is there 3 times the same DELETE query ?

File: src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/ImportDataConfigurationController.php
Comment (matks) [inline src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/ImportDataConfigurationController.php]: Please do not use `@Template` annotation ;) as other controllers use `$this->render()`

File: src/PrestaShopBundle/Resources/config/routing/admin/configure/advanced_parameters/import.yml
Comment (matks) [inline src/PrestaShopBundle/Resources/config/routing/admin/configure/advanced_parameters/import.yml:9]: 👍

File: tests/Unit/Core/Image/Deleter/ImageFileDeleterTest.php
Comment (matks) [inline tests/Unit/Core/Image/Deleter/ImageFileDeleterTest.php:35]: 👏

**Review by matks — CHANGES_REQUESTED** [review]

**Comment by matks** [comment]
Hi @rokaszygmantas ! Code review is done. Sorry it was late, computer crashed on Friday 😬 

While reviewing I understood a lot of things:
1) Thank you for handling the migration of Import feature, it is VERY complex and hard to read. Must have been tough for you.
2) The Import feature is complex and has a lot of bad code in it. RIght now we are migrating it but I can see that this is a flaw for PrestaShop. It would require a complete re-write one day (not today, and not for you though). For example there are several logic statements which are duplicated (how to create a product, how to create a category, how do images are being stored, SQL table names ...) although it should be functions reused between the Import page and the CRUD pages (there should be a single way to create a category, reused everywhere). Currently, if we modify one of these be items in PrestaShop the Import process will crash.
3) So we are going to stay with this implementation for a while. In order to protect the field for future work, we'll need massive integration testing. See below.
4) I reviewed the PR but it's very complex so I only mentioned small issues. The task is too big, for me and you, for a thorough review.

## Integration testing

In order to allow a future rework, I would like that we setup a dedicated test folder in `tests/Integration`. We can call it simply `tests/Integration/Import`. In this folder, we must use one testing system (can be Behat, can be Phpunit, can be anything in fact) that will do the following:
- boot/setup required PrestaShop items for the Import feature to work (might be the whole SF application if needed or if it makes things simpler, see https://github.com/PrestaShop/PrestaShop/pull/11560 for example)
- take a bunch of CSV import samples, and for each of them perform the import and check the output

So this would be "black box" tests: setup system, set an initial state, apply change (here, a CSV import being processed), check final test, reset initial state, repeat with next sample.

What do you think about it ?

File: src/Adapter/Import/ImportEntityDeleter.php
Comment (rokaszygmantas) [reply src/Adapter/Import/ImportEntityDeleter.php:176]: `category`, `category_lang`, `category_shop` tables :slightly_smiling_face:

Comment (matks) [reply src/Adapter/Import/ImportEntityDeleter.php:176]: Argh sorry 😅 at this point in the PR my brain was kindof tired 😢

**Comment by rokaszygmantas** [comment]
I agree with the idea of integration tests :) added it to the list.

File: src/Adapter/Import/Handler/AbstractImportHandler.php
Comment (rokaszygmantas) [reply src/Adapter/Import/Handler/AbstractImportHandler.php:237]: yeah https://github.com/PrestaShop/PrestaShop/blob/develop/controllers/admin/AdminImportController.php#L4607

File: src/Core/Import/Handler/ImportHandlerInterface.php
Comment (rokaszygmantas) [reply src/Core/Import/Handler/ImportHandlerInterface.php:45]: All imports need these methods at the moment, but only because it's needed for legacy logic to be executed. 

Anyway, I still think it's good to have those methods in the core interface because if you check this interface, it has an `importRow()` method, so that method is called once for every row that's being imported. `setUp()` and `tearDown()` are called once per whole import process, so it can be used to run hooks in the future, write information to logs, send emails or something else that's needed to be executed only once.

wdyt @matks?

File: src/Adapter/Import/Handler/CategoryImportHandler.php
Comment (rokaszygmantas) [reply src/Adapter/Import/Handler/CategoryImportHandler.php:386]: yeah that definitely needs to be unified, but I think it's risky at the moment to modify this logic. The whole import logic needs a huge overhaul in the future :slightly_smiling_face:

File: src/Adapter/Import/Handler/CategoryImportHandler.php
Comment (rokaszygmantas) [reply src/Adapter/Import/Handler/CategoryImportHandler.php:52]: yes, have you seen my checklist? :smile: I've completed products and categories logic migration to adapters, now combinations logic is WIP. The PR is getting bigger and bigger, but I don't think there is a good way to split it, so it has to be done here. :slightly_smiling_face:

File: src/Adapter/Import/Handler/ProductImportHandler.php
Comment (rokaszygmantas) [reply src/Adapter/Import/Handler/ProductImportHandler.php:327]: yeah this was painful to handle :smile: but anyhow, that's an adapter - it's meant to be removed in the future.
https://github.com/PrestaShop/PrestaShop/blob/develop/controllers/admin/AdminImportController.php#L1865 (notice `$this` as an argument)

File: src/Adapter/Import/Handler/CategoryImportHandler.php
Comment (matks) [reply src/Adapter/Import/Handler/CategoryImportHandler.php:52]: OK then please just tell me the commits sha1 I have to look at ;) so I know what has been reviewed and what has not been reviewed yet

File: src/PrestaShopBundle/Resources/config/services/adapter/data_provider.yml
Comment (rokaszygmantas) [reply src/PrestaShopBundle/Resources/config/services/adapter/data_provider.yml:112]: I think a good way could be to create a new employee data provider implementation for non-BO cases when it's needed - there's a `ContextEmployeeProviderInterface` for that reason :slightly_smiling_face:

File: src/Adapter/Import/CsvFileOpener.php
Comment (rokaszygmantas) [reply src/Adapter/Import/CsvFileOpener.php:94]: It's a legacy logic, but I think if we don't `rewind()` the second time, the file pointer would be at 3rd position when the file doesn't contain BOM, meaning it would skip 3 characters from the beginning of the file. :thinking: because `fread($handle, 3)` already moves the file pointer.

File: src/Adapter/Import/Handler/ImportHandlerFinder.php
Comment (rokaszygmantas) [reply src/Adapter/Import/Handler/ImportHandlerFinder.php]: oh yes this is a very good idea :)

**Comment by mickaelandrieu** [comment]
Hello,

I advise you to split this big tasks into multiple pull requests or the risk will be that no one will be easy to merge it.

How about split each checkbox to a dedicated issue and pull request?

Mickaël

**Comment by rokaszygmantas** [comment]
Hi @mickaelandrieu, thanks for suggestion! 

I would really like to split it, because the PR is already enormous big. But should I wait for each PR to be merged before opening another, or what could be the strategy? 
The problem is that a lot of the code would need to be shared across these pull requests.

**Comment by rokaszygmantas** [comment]
@matks I made some fixes starting from commit `b18683c` :)

**Review by matks — DISMISSED** [review]

**Comment by matks** [comment]
OK, we're going to merge this PR and start another
- up-to-date code review 👍 ✅ 
- need to rebase first
- then QA should check it works 😉

**Comment by rokaszygmantas** [comment]
rebased.

**Review by matks — DISMISSED** [review]

**Comment by rokaszygmantas** [comment]
btw assets are not rebuilt in this PR, but I was told that it's no longer needed to include built assets in PRs :)

**Comment by matks** [comment]
@rokaszygmantas yes indeed
QA team can build them, however we need to tell them in the PR description (done)
and we'll build them only after we got their approval (to avoid this PR to be blocked by a assets rebase)

**Comment by rokaszygmantas** [comment]
thanks, I will include it in the description in future PRs :)

**Review by matks — DISMISSED** [review]

**Comment by rokaszygmantas** [comment]
@matks apparently after rebase the import is throwing a fatal error when visiting it. The same happens in develop branch. I'll try to open a new PR with a fix, maybe we should remove the `waiting for QA` label from this PR while it's not fixed.

**Comment by matks** [comment]
@rokaszygmantas Yes indeed 👍 
This is probably because of modified files inside `1.7.5.x` that have been merged later into `develop`

**Comment by marionf** [comment]
I remove waiting for QA label, please put it back when #11753 will be fixed :)

**Comment by rokaszygmantas** [comment]
@matks fixed the merge errors and rebased :)

**Comment by matks** [comment]
@rokaszygmantas We have Travis complaining about lint 😢 you must apply php-cs-fixer fixes:
```
vendor/bin/php-cs-fixer fix
```

**Comment by rokaszygmantas** [comment]
@matks rebased once again and ran the cs fixer :smile:

**Comment by matks** [comment]
👏

**Review by matks — DISMISSED** [review]

**Review by matks — DISMISSED** [review]

**Comment by marionf** [comment]
Hello @rokaszygmantas 

1) Some columns was matching before and are not now

capture d ecran_812

2) I have differents results with the same CSV between the legacy page and the symfony page

**For products**

capture d ecran_811

The result is: with the legacy 0/3 products imported with differents errors, with the symfony 1/3 without any errors

capture d ecran_814

This is the product CSV used: 
https://drive.google.com/file/d/10y6I_x-fzETokDDCphX-zBqGMEwBpyG-/view?usp=sharing

**For categories**

capture d ecran_815

The result is: with the legacy 3/3 categories imported with differents errors, with the symfony 0/3 without the same errors as in legacy

capture d ecran_816

If I try to ignore the column link rewrite, I have the error below (I don't have it in legacy)

capture d ecran_817

This is the category CSV used:
https://drive.google.com/file/d/1dzxFunS7H17Esmi6mflxy4nKMxCKbhtF/view?usp=sharing

**Comment by rokaszygmantas** [comment]
hi @matks, as we've discussed on gitter - I've restored the legacy import page 2 now. Migrated page 2 is hidden now, but it's code still exists. I think we need another QA round on this, wdyt?

**Comment by matks** [comment]
@marionf After discussing with @colinegin, we decided to put the migration of this page on "pause" because it is so complex and long. We'll come back on it for 1.7.7 .

Can you please check that we have successfully hidden the work-in-progress and restored the 1..5 behavior (Import page 1 migrated, Import page 2 legacy) ?

**Review by matks — DISMISSED** [review]

**Comment by PierreRambaud** [comment]
Still work in progress?

**Comment by rokaszygmantas** [comment]
@PierreRambaud no, as per @matks comment above, the import migration will be paused, so this PR should be reviewed and merged. :slightly_smiling_face: 
just noticed that travis fails though, I'll need to run the code sniffer. :sweat_smile:

**Comment by rokaszygmantas** [comment]
@matks, I've done another rebase and ran the CS fixer as well, so your review got dismissed. :slightly_smiling_face:

**Review by matks — APPROVED** [review]

**Comment by matks** [comment]
@rokaszygmantas got it, finally 😄

**Comment by matks** [comment]
Thanks @rokaszygmantas

**Pull request merged**
